pubsub/google/cloud/pubsub_v1/subscriber/_protocol/bidi.py

```diff
@@ -421,16 +421,25 @@ def _thread_main(self):
             self._bidi_rpc.open()
 
             while self._bidi_rpc.is_active:
-                if not self.is_paused:
+                # Do not allow the paused status to change at all during this
+                # section. There is a condition where we could be resumed
+                # between checking if we are paused and calling wake.wait(),
+                # which means that we will miss the notification to wake up
+                # (oops!) and wait for a notification that will never come.
+                # Keeping the lock throughout avoids that.
+                self._wake.acquire()
+
+                if not self._paused:
+                    self._wake.release()
                     _LOGGER.debug('waiting for recv.')
                     response = self._bidi_rpc.recv()
                     _LOGGER.debug('recved response.')
                     self._on_response(response)
                 else:
                     _LOGGER.debug('paused, waiting for waking.')
-                    with self._wake:
-                        self._wake.wait()
+                    self._wake.wait()
                     _LOGGER.debug('woken.')
+                    self._wake.release()
```

Comment:
ISTM that the locking would be clearer if it used `self._wait` as a context manager for all the protected sections, which likely means inverting the condition.  E.g.:

```python
            while self._bidi_rpc.is_active:

                with self._wake:

                    paused = self._paused:

                    if paused:
                        _LOGGER.debug('paused, waiting for waking.')
                        self._wake.wait()
                        _LOGGER.debug('woken.')

                if not paused:
                    _LOGGER.debug('waiting for recv.')
                    response = self._bidi_rpc.recv()
                    _LOGGER.debug('recved response.')
                    self._on_response(response)
```

Reply:
Great suggestion, I'll do this.

Reply:
Gah, if we could drop 2.7 we could use https://docs.python.org/3/library/threading.html#threading.Condition.wait_for